### PR TITLE
FileSystemAsyncUploadStorage and other fixes

### PR DESCRIPTION
**Pull request by davecoates**
Add `FileSystemAsyncUploadStorage` for uploading direct to local filesystem. Useful in local dev, or for faster experience on forms (files can start uploading immediately while form still being filled out)

Add migration guide for legacy projects

Various other tweaks and fixes from testing in projects.

### CODE REVIEW

File: packages/ap-storage/alliance_platform/storage/filesystem_views.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            return HttpResponseForbidden("Invalid upload URL.")
+
+        if "file" not in request.FILES:
+            return HttpResponseBadRequest("No file uploaded.")
+
+        # Save the uploaded file to the specified path
+        uploaded_file = cast(UploadedFile, request.FILES["file"])
+        storage.save(path, uploaded_file)
+
+        return HttpResponse("File uploaded successfully.")
```
Comment (micahklaz) [inline packages/ap-storage/alliance_platform/storage/filesystem_views.py:81]: Should this be a 201 response?

File: packages/ap-storage/docs/index.rst [hunk 2]
```diff
@@ -1,13 +1,16 @@
 Alliance Platform Storage
 =============================================
 
-This package provides storage classes & fields to make dealing with storage providers easier (e.g. S3, Azure).
+This package provides classes & fields for working with storage easier.
```
Comment (micahklaz) [inline packages/ap-storage/docs/index.rst:4]: Suggestion: easier -> more easily

File: packages/ap-storage/alliance_platform/storage/filesystem.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+from alliance_platform.storage.registry import AsyncFieldRegistry
+from django.core.files.move import file_move_safe
+from django.core.files.storage import FileSystemStorage
+from django.core.signing import TimestampSigner
+from django.urls import path
+from django.urls import reverse
+
+
+class FileSystemAsyncUploadStorage(FileSystemStorage, AsyncUploadStorage):
+    """Implementation of AsyncUploadStorage that uploads to direct the local server
```
Comment (micahklaz) [inline packages/ap-storage/alliance_platform/storage/filesystem.py:15]: "to direct the" -> "directly to the"

**Review by micahklaz — CHANGES_REQUESTED** [review]

File: packages/ap-storage/alliance_platform/storage/fields/async_file.py [hunk 4]
```diff
@@ -482,10 +482,13 @@ class AsyncFileInput(Input):
     of :class:`~alliance_platform.storage.fields.async_file.AsyncFileInputData`. This is then handled
     on the descriptor classes for :class:`~alliance_platform.storage.fields.async_file.AsyncFileField`
     and :class:`~alliance_platform.storage.fields.async_file.AsyncImageField`.
+
+    To customise the widget rendered on the frontend you can override the template
+    ``alliance_platform/storage/widgets/async_file_input.html``.
```
[the remark this answers is not in the document]
Comment (davecoates) [reply packages/ap-storage/alliance_platform/storage/fields/async_file.py:487]: Eventually, once I finish the UI widget (soon). For now it has to be provided in the project

File: packages/ap-storage/docs/async_uploads.rst [hunk 5]
```diff
@@ -0,0 +1,229 @@
+Async File Uploads
+==================
+
+Overview
+--------
+
+The Async Upload feature in alliance-platform-storage provides a seamless way to handle asynchronous file uploads in Django applications.
+In this context, the async in "Async Upload" means the file is immediately uploaded to the server (e.g. S3, Azure, local server) while the user
+continues to fill out a form. A reference (the key) to the uploaded file is stored in the form and handled in the Django model save when
+processing the form submission. See `How it works`_ for more specifics.
```
Comment (micahklaz) [inline packages/ap-storage/docs/async_uploads.rst:10]: This is great - it took me quite a while to originally figure out what the actual point of the async upload widget was, and this explains it very well.

File: packages/ap-storage/docs/async_uploads.rst [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+5. Upon form submission, the backend moves the file from its temporary location to its final destination, and cleans up the :class:`~alliance_platform.storage.models.AsyncTempFile` record.
+
+6. If form submission never occurs, for example the user abandons the form after uploading a file, then the file will be retained until
+   the :djmanage:`cleanup_async_temp_files` command is run.
+
+Usage
+-----
+
+1. Add a :class:`~alliance_platform.storage.fields.async_file.AsyncFileField`: or :class:`~alliance_platform.storage.fields.async_file.AsyncImageField` to a model, optionally passing
+   the ``storage`` option depending on what the :setting:`STORAGES <django:STORAGES>`
```
Comment (micahklaz) [inline packages/ap-storage/docs/async_uploads.rst:174]: Maybe change this line to  something like

the ``storage`` option if you need to use a different backend from the project :setting:`STORAGE <django:STORAGES>` setting

**Review by micahklaz — CHANGES_REQUESTED** [review]

File: packages/ap-storage/docs/async_uploads.rst (hunk 6 above)
Comment (davecoates) [reply packages/ap-storage/docs/async_uploads.rst]: thanks, good one - fixed

**Review by micahklaz — APPROVED** [review]

**Pull request merged**